Mathlib/Tactic/LibrarySearch.lean

```diff
@@ -46,23 +47,58 @@ inductive DeclMod
 | none | symm | mp | mpr
 deriving DecidableEq
 
-initialize librarySearchLemmas : DeclCache (DiscrTree (Name × DeclMod) true) ←
-  DeclCache.mk "librarySearch: init cache" {} fun name constInfo lemmas => do
-    if constInfo.isUnsafe then return lemmas
-    if ← name.isBlackListed then return lemmas
-    withNewMCtxDepth do withReducible do
-      let (_, _, type) ← forallMetaTelescopeReducing constInfo.type
-      let keys ← DiscrTree.mkPath type
-      let lemmas := lemmas.insertIfSpecific keys (name, .none)
-      match type.getAppFnArgs with
-      | (``Eq, #[_, lhs, rhs]) => do
-        let keys_symm ← DiscrTree.mkPath (← mkEq rhs lhs)
-        pure (lemmas.insertIfSpecific keys_symm (name, .symm))
-      | (``Iff, #[lhs, rhs]) => do
-        let keys_mp ← DiscrTree.mkPath rhs
-        let keys_mpr ← DiscrTree.mkPath lhs
-        pure <| (lemmas.insertIfSpecific keys_mp (name, .mp)).insertIfSpecific keys_mpr (name, .mpr)
-      | _ => pure lemmas
+/-- Insert a lemma into the discrimination tree. -/
+def addLemma (name : Name) (constInfo : ConstantInfo)
+    (lemmas : DiscrTree (Name × DeclMod) true) : MetaM (DiscrTree (Name × DeclMod) true) := do
+  if constInfo.isUnsafe then return lemmas
+  if ← name.isBlackListed then return lemmas
+  withNewMCtxDepth do withReducible do
+    let (_, _, type) ← forallMetaTelescopeReducing constInfo.type
+    let keys ← DiscrTree.mkPath type
+    let lemmas := lemmas.insertIfSpecific keys (name, .none)
+    match type.getAppFnArgs with
+    | (``Eq, #[_, lhs, rhs]) => do
+      let keys_symm ← DiscrTree.mkPath (← mkEq rhs lhs)
+      pure (lemmas.insertIfSpecific keys_symm (name, .symm))
+    | (``Iff, #[lhs, rhs]) => do
+      let keys_mp ← DiscrTree.mkPath rhs
+      let keys_mpr ← DiscrTree.mkPath lhs
+      pure <| (lemmas.insertIfSpecific keys_mp (name, .mp)).insertIfSpecific keys_mpr (name, .mpr)
+    | _ => pure lemmas
+
+/-- Construct the discrimination tree of all lemmas. -/
+def buildDiscrTree : IO (DeclCache (DiscrTree (Name × DeclMod) true)) :=
+  DeclCache.mk "librarySearch: init cache" {} addLemma
+
+open System (FilePath)
+
+def cachePath : IO FilePath :=
+  try
+    return (← findOLean `MathlibExtras.LibrarySearch).withExtension "extra"
```

Comment:
there is a function `getMathlibDir` for this

Reply:
I don't understand why I'd use that here: I'm looking to put a `.extra` file beside the `.olean` file, so `Cache` can handle it.